File: C++/unique-morse-code-words.cpp
// Time:  O(n), n is the sume of all word lengths
// Space: O(n)

class Solution {
public:
    int uniqueMorseRepresentations(vector<string>& words) {
        static const vector<string> MORSE = 
            { ".-", "-...", "-.-.", "-..", ".", "..-.", "--.",
              "....", "..", ".---", "-.-", ".-..", "--", "-.",
              "---", ".--.", "--.-", ".-.", "...", "-", "..-",
              "...-", ".--", "-..-", "-.--", "--.."};

        unordered_set<string> lookup;
        for (const auto& word : words) {
            string code;
            for (const auto& c : word) {
                code += MORSE[c - 'a'];
            }
            lookup.emplace(move(code));
        }
        return lookup.size();
    }
};

File: Python/Unique-Morse-Code-Words.py
class Solution:
    def uniqueMorseRepresentations(self, words: List[str]) -> int:        
        result = set()
        def to_morse(s):
            return ''.join(mapp.get(i) for i in s)
        
        mapp = {'a': '.-',     'b': '-...',   'c': '-.-.', 
                'd': '-..',    'e': '.',      'f': '..-.',
                'g': '--.',    'h': '....',   'i': '..',
                'j': '.---',   'k': '-.-',    'l': '.-..',
                'm': '--',     'n': '-.',     'o': '---',
                'p': '.--.',   'q': '--.-',   'r': '.-.',
                's': '...',    't': '-',      'u': '..-',
                'v': '...-',   'w': '.--',    'x': '-..-',
                'y': '-.--',   'z': '--..',
                }
        
        
        for word in words:
            result.add(to_morse(word))
            
        return len(result)
        
